Pm rounded button

pm-dimagi opened the pull request:
This PR is related to the design changes and it contains the update of the button all over the app now the square button is converted to the rounded button

## Feature Flag
https://dimagi.atlassian.net/browse/CCCT-536

## Release Note
All the buttons in the commcare side is converted to the rounded button as per the figma design 
link to the figma is attached [[here](https://www.figma.com/design/HntzHo84kM9084Nz78567m/Dimaagi-Project?node-id=4979-13873&t=MyufA7vYbANQhvtg-4)](url)

## Product Description
Converted all the buttons to the rounded buttons
![WhatsApp Image 2024-11-15 at 01 45 38 (1)](https://github.com/user-attachments/assets/2d267419-7a66-4b19-91f6-0b0850a0b969)
![WhatsApp Image 2024-11-15 at 01 45 38 (2)](https://github.com/user-attachments/assets/b75c3056-0659-4234-b178-477ce3a06b6c)
![WhatsApp Image 2024-11-15 at 01 45 38](https://github.com/user-attachments/assets/d54cfae1-f5e1-4728-910f-c2754b821e37)
![clear](https://github.com/user-attachments/assets/d8fce12d-3e02-4b8b-8a2e-93fe3653d2a9)
![WhatsApp Image 2024-11-27 at 14 48 42](https://github.com/user-attachments/assets/8ce37419-b195-4fc6-bb18-8f5d0769e61b)
![WhatsApp Image 2024-11-28 at 18 48 19 (1)](https://github.com/user-attachments/assets/eb932007-7144-4f3a-a081-ed01a0cbce9a)
![WhatsApp Image 2024-11-28 at 19 34 40](https://github.com/user-attachments/assets/e2628e63-dcec-4a7e-9d89-8180139d6a84)

![WhatsApp Image 2024-11-28 at 18 49 20](https://github.com/user-attachments/assets/d14899b5-aa60-4657-a6dc-efbc38e47283)

Doc for the screenshot reference and requested changes
https://docs.google.com/document/d/1xkqPrLClFrUONOa6gIOk1408NE1S23Wjy-cCefIGdUI/edit?usp=sharing


https://github.com/user-attachments/assets/3485f123-1205-488d-bbf3-d640b8d7c873


https://github.com/user-attachments/assets/5f807e1e-6e38-4270-babc-00b94ebc7173


https://github.com/user-attachments/assets/7f5fddad-98c0-4533-9654-6852ad986009


https://github.com/user-attachments/assets/bb7e5cf7-c82d-4655-b30f-38b2497c8887



## PR Checklist

- [x] If I think the PR is high risk, "High Risk" label is set
- [x] I have confidence that this PR will not introduce a regression for the reasons below
- [x] Do we need to enhance manual QA test coverage ? If yes, "QA Note" label is set correctly
- [x] Does the PR introduce any major changes worth communicating ? If yes, "Release Note" label is set and a "Release Note" is specified in PR description.

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->


### Safety story
I have tested the one of the forms with app code 48PZ1dJ
I have tested Login flow
I have tested settings page
I have tested menu options  

QA Note: All buttons during normal testing should be carefully examined to observe any anamolies or anything that doesn't seem to fit well in overall UX of the screen. 

Release Note: Change in Button Styling.

avazirna commented on app/res/layout-land/activity_crash_warning.xml:
```diff
@@ -93,12 +93,11 @@
 
-    <Button
+    <com.google.android.material.button.MaterialButton
+        style="@style/CustomButtonStyle"
```
It's better to apply the button style globally in the app's theme and only set it to specific buttons when a different style is needed

avazirna commented on app/res/layout/activity_drawing_boundary.xml:
```diff
@@ -44,32 +44,36 @@
         android:orientation="horizontal">
 
-        <Button
-            style="@style/MapButtonGreen"
```
This button had a specific style, why change it to the _default_?

avazirna commented on app/res/values/styles.xml:
```diff
@@ -305,4 +305,32 @@
+    <style name="CustomButtonStyleOutline" parent="Widget.MaterialComponents.Button.OutlinedButton">
+        <item name="shapeAppearanceOverlay">@style/MyShapeAppearance</item>
+    </style>
+    <style name="MyShapeAppearance" parent="">
```
_MyShapeAppearance_ is too generic, better to choose a more context related name

avazirna reviewed (commented):
@pm-dimagi a couples of questions:
1. I noticed that you applied the `CustomButtonStyle` to buttons that had specific styles (different from the _default_ style), did you test the affected features to validate the visual impact?
2. Here is the link too the figma https://www.figma.com/design/UWjp1wRNAO0o9q2llwDKMc/CommCare-Cosmetic-Changes-Post-Nov-'22Rebrand?node-id=111-2&node-type=canvas&t=i0Tr9NVN0YY4o3SC-0
3. This is a silly one, but I'm not sure about the rounded corners `DONE` button. I think because it spans the entire width of the parent view, it doesn't give a good visual effect. Did that design come form the marketing team?
![image](https://github.com/user-attachments/assets/3e7457fc-4867-4dfa-bc16-3b49c9113c0d)

pm-dimagi commented on app/res/values/colors.xml:
```diff
@@ -146,4 +146,9 @@
     <color name="button_disabled">#d0e3ff</color>
+
+
+    <color name="button_primary_color"> #3A42C7</color>
```
it can be used at some other places also thats y

shubham1g5 commented:
@pm-dimagi This should not include changes from https://github.com/dimagi/commcare-android/pull/2892 to minimise the reviewer effort, as such you should change the base branch here to the other PR instead of `master`.

shubham1g5 commented:
@pm-dimagi An essential part of creating PR is to ensure all  PR Checklist boxes being thought of and ticked. (You should tick all items of the checklist to communicate that you have thought and taken proper action on each of the items) 

Also wants to flag that your PR ticks the `Does the PR introduce any major changes worth communicating ? If yes, "Release Note" label is set and a "Release Note" is specified in PR description.` item without really specifying the "Release Note" in the PR description .

shubham1g5 commented on app/res/layout/activity_drawing_boundary.xml:
```diff
@@ -44,32 +44,32 @@
-        <Button
-            style="@style/MapButtonGreen"
+        <com.google.android.material.button.MaterialButton
+            android:backgroundTint="@color/green_500"
```
The old styled defined a selector that changes based on pressed button state, any reason we are no longer doing that ?

pm-dimagi replied:
This is updated in the last commit i have created the new style for this which is of same behaviour as the previous one

shubham1g5 replied:
can you please link me to the style you are referring to ?

shubham1g5 commented on app/res/layout/blue_outlined_button.xml:
```diff
@@ -1,13 +1,14 @@
 <?xml version="1.0" encoding="utf-8"?>
-<Button xmlns:android="http://schemas.android.com/apk/res/android"
+<com.google.android.material.button.MaterialButton xmlns:android="http://schemas.android.com/apk/res/android"
```
can we have a screenshot of this button in PR description as well, it's the 'Clear' button that appears on configuing a [multi-select question in the form](https://dimagi.atlassian.net/wiki/spaces/commcarepublic/pages/2213544088/List+of+Question+Types+in+CommCare#Multiple-Choice)

pm-dimagi replied:
![clear](https://github.com/user-attachments/assets/002509ff-f5aa-431a-b9c8-d8ff4a08ef38)

shubham1g5 commented on app/res/layout/choice_dialog_view.xml:
```diff
@@ -25,19 +25,19 @@
     </ListView>
 
-    <Button xmlns:android="http://schemas.android.com/apk/res/android"
+    <com.google.android.material.button.MaterialButton
```
Also requesting a screenshot of the choice dialog with buttons in product description. (You can see it when configuing a [repeat group](https://dimagi.atlassian.net/wiki/spaces/commcarepublic/pages/2143956644/Repeat+Groups) in the form)

avazirna replied:
Here's an example from the `Exit form` dialog:
![exit_form_dialog_new_ui](https://github.com/user-attachments/assets/69392efb-963d-4043-847f-94deb3b7ee1f)
I don't think these need to be rounded corners buttons, here's the current design:
![image](https://github.com/user-attachments/assets/0911f79a-ac89-4fba-bf79-2b5c2b30fa46)

shubham1g5 reviewed (changes requested):
@pm-dimagi  I started to review it but stopped in between as I think there are various places where the button's existing style has been removed, this would be impractical to review on code level without either having screenshots of the impacted place Or a reasoning on why you think removing old style would not cause non-intented changes with that specific button OR a comment in Safety story mentioning that you have made sure that the button doesn't appear in some weird fashion for it's respective usage on UI. If you have trouble figuiring out where a specific xml layout is used feel free to ping the mobile slack. 

Regarding the safety story, I think you will need to test the app out at several places and all different widgets. I good way to do that is to run the instrumentations tests locally on your phone, specifically the [FormEntryTest](https://github.com/dimagi/commcare-android/blob/master/app/instrumentation-tests/src/org/commcare/androidTests/FormEntryTest.kt)

Also it would be great if you can respond to Ahmad's [comment above](https://github.com/dimagi/commcare-android/pull/2897#pullrequestreview-2442711714) here as I am curious about the points raised by Ahmad as well.

pm-dimagi commented:
@shubham1g5 @avazirna this is to confirm that  I have removed the old styles of the buttons as I have to give the new style of rounded shape but I have made sure that the other properties of that style is incoperatesd in the xml itself like for the MapButtonGreen I have given the green color speciic to the Material button instead of passing through the style.
The done button was told not looking good for that particular ui I Have reverted that change

avazirna commented on app/res/layout/hierarchy_layout.xml:
```diff
@@ -40,40 +40,31 @@ the License.
-            android:background="@color/cc_core_bg"
-            style="@style/CommCare.Button.Default"
             android:layout_weight="1"/>
 
-        <Button
+        <com.google.android.material.button.MaterialButton
             android:id="@+id/jumpBeginningButton"
             android:layout_width="wrap_content"
             android:layout_height="wrap_content"
             android:text="@string/jump_to_beginning"
             android:padding="12dp"
             android:textSize="@dimen/text_medium"
-            android:textColor="@color/cc_brand_color"
-            android:background="@color/cc_core_bg"
-            style="@style/CommCare.Button.Default"
             android:layout_weight="1"/>
 
-        <Button
+        <com.google.android.material.button.MaterialButton
             android:id="@+id/jumpEndButton"
             android:layout_width="wrap_content"
             android:layout_height="wrap_content"
             android:text="@string/jump_to_end"
             android:padding="12dp"
             android:textSize="@dimen/text_medium"
-            android:textColor="@color/cc_brand_color"
-            android:background="@color/cc_core_bg"
-            style="@style/CommCare.Button.Default"
             android:layout_weight="1"/>
```
I think that visually this looks fine but there are a few aspects to address:
(1) when disabled, the contrast between the background and the text makes the label unreadable
(2) there is space between the left edge and the first button, making it look misaligned
(3) maybe we should capitalize all the letters as it is now
![image](https://github.com/user-attachments/assets/a1ae84c7-a7d7-4b4d-b59e-87eb9a34bc00)
this is how it looks now:
![image](https://github.com/user-attachments/assets/4158feea-12e4-42b7-bcd3-5984f6d966f5)

pm-dimagi commented:
@damagatchi retest this please

shubham1g5 commented on app/res/values/styles.xml:
```diff
@@ -305,4 +305,32 @@
     </style>
+
+        <style name="CustomButtonStyle" parent="Widget.MaterialComponents.Button">
+            <item name="background">@color/start_sync_button</item>
```
should be named appropriately like - `button_background_color`

shubham1g5 commented on app/res/layout/single_dialog_choice_view.xml:
```diff
@@ -1,11 +1,12 @@
 <?xml version="1.0" encoding="utf-8"?>
-<Button xmlns:android="http://schemas.android.com/apk/res/android"
+<Button
```
Why was this not changed ?

pm-dimagi replied:
Because of exit form UI

shubham1g5 commented on app/res/layout/scrolling_calendar_widget.xml:
```diff
@@ -134,12 +134,10 @@
             android:layout_alignParentStart="true" />
 
-        <Button
+        <com.google.android.material.button.MaterialButton
```
we need a screenshot to review. This can be accessed by [configuring a calendar on a date widget](https://dimagi.atlassian.net/wiki/spaces/commcarepublic/pages/2143957035/Date+and+Calendar+Display#Use-of-the-Alternate-Calendar-Systems-in-CommCare-Android).

pm-dimagi replied:
![calender ss](https://github.com/user-attachments/assets/08938797-2fe2-49e6-ac36-46a8667a4acb)

shubham1g5 replied:
The [calendar button](https://github.com/dimagi/commcare-android/blob/master/app/res/drawable/icon_calendar_circle_turquoise.xml) only appears for the Gregorian widget, you will need to set the appearance attribute as `gregorian` on your date widget to be able to see it.

shubham1g5 reviewed (changes requested):
Can we also have a few screenshots on a different form factor like a tablet ?

pm-dimagi commented on app/res/layout/hierarchy_layout.xml:
```diff
@@ -46,10 +46,10 @@ the License.
+            android:textSize="@dimen/text_medium"
             android:layout_weight="1"/>
 
         <Button
```
![image (3)](https://github.com/user-attachments/assets/abbf01f2-9fdc-45b1-bda5-1ef903918284)
Because ahmad raised the issue in rounded design at the top thats why i converted into simple button

pm-dimagi commented on app/res/layout/connection_diagnostic.xml:
```diff
@@ -39,10 +39,10 @@
             android:paddingStart="5dp"
             android:paddingEnd="5dp">
 
-            <Button
```
adding at top

pm-dimagi commented on app/res/layout/entity_detail.xml:
```diff
@@ -9,8 +10,12 @@
     tools:viewBindingIgnore="true">
 
     <Button
```
Because the top finish button andd done button was coming rounded and was not looking good so ahmad and I decides to leave as it is

shubham1g5 reviewed (approved):
Good to go!

shubham1g5 commented:
@pm-dimagi Seems like rabbit caught some critical issues with the changes here, can you have a pass on code rabbit comments over there and address what you think is vital to address.

avazirna commented:
> @pm-dimagi Seems like rabbit caught some critical issues with the changes here, can you have a pass on code rabbit comments over there and address what you think is vital to address.

Looking into these now

avazirna commented on app/src/org/commcare/CommCareNoficationManager.java:
```diff
@@ -77,7 +77,7 @@ private void updateMessageNotification() {
                         .setContentTitle(title)
                         .setContentText(Localization.get("notifications.prompt.details", new String[]{additional}))
-                        .setSmallIcon(R.drawable.notification)
+                        .setSmallIcon(R.drawable.commcare_actionbar_logo)
```
I have pushed a fix in commit 9365782329890b9d995eb772d55b2b096049d788, please review it.

shubham1g5 commented on app/res/values/styles.xml:
```diff
@@ -328,7 +328,7 @@
         <item name="shapeAppearanceOverlay">@style/RoundedButtonStyle</item>
     </style>
-    <style name="RoundedButtonStyle" parent="">
+    <style name="RoundedButtonStyle" parent="Widget.MaterialComponents.Button.OutlinedButton">
```
Are you sure that `Widget.MaterialComponents.Button.OutlinedButton` is the right style to be applied ? And Have you tested impact of changing the parent Style here ?

avazirna replied:
@shubham1g5 this is used to set the `shapeAppearanceOverlay` of the `CustomButtonStyleOutline` style so it should be fine. And yes, I tested it and it seemed fine, applied to the _CANCEL_ button below:
![image](https://github.com/user-attachments/assets/b33cf853-67cf-4a45-a02a-954ac8042b20)

shubham1g5 replied:
ahh can we rename this style to suggest that it's only applied for alert dialogs. `RoundedButtonStyle` seems too generic if it's only for dialogs.

shubham1g5 replied:
curious why do we need to apply it to the clear button  ?

avazirna replied:
I have no context about that decision, it's the same style as before. I believe it's considered a negative action

shubham1g5 commented on app/res/layout/blue_outlined_button.xml:
```diff
@@ -1,13 +1,13 @@
     android:id="@+id/blue_outlined_button"
     android:layout_width="wrap_content"
     android:layout_height="wrap_content"
     android:layout_gravity="right"
```
@avazirna this would be nice to correct.

shubham1g5 commented on app/res/layout/pin_screen_buttons.xml:
```diff
@@ -8,14 +8,14 @@
     tools:viewBindingIgnore="true">
 
-    <Button
+    <com.google.android.material.button.MaterialButton
         android:layout_width="wrap_content"
         android:layout_height="@dimen/min_button_height"
         android:id="@+id/pin_cancel_button"
         android:textSize="@dimen/font_size_dp_medium"
```
@avazirna Wanted to ask if it's intentional to not have the style applied here and if we have verified the pin screen in our testing.

avazirna replied:
@shubham1g5 we applied the style to all material buttons globally [here](https://github.com/dimagi/commcare-android/pull/2897/files#diff-d14add8196b9ab34379ce4d7c25b63972dc909be3097bebd6d74b7d2043094d6R12), but I didn't test the PIN screen, let me do that

shubham1g5 commented on app/res/layout/screen_form_dump.xml:
```diff
@@ -20,9 +20,9 @@
             android:scaleType="fitStart"
-            android:maxHeight="135px"
+            android:maxHeight="125px"
             android:maxWidth="663px"
-            app:srcCompat="@drawable/ccbanner"
+            app:srcCompat="@drawable/commcare_by_dimagi"
```
@avazirna bump.

shubham1g5 commented on app/res/values/styles.xml:
```diff
@@ -305,4 +305,33 @@
+            <item name="iconPadding">8dp</item>
+            <item name="android:textAllCaps">false</item>
+            <item name="android:textStyle">bold</item>
+        </style>
```
@avazirna +1 to this.

avazirna replied:
@shubham1g5 I had done this already in this commit: https://github.com/dimagi/commcare-android/pull/2897/commits/3dcb2f1c3ba7693998db53ec40c9282c3e756100

avazirna commented on app/res/values/styles.xml:
```diff
@@ -325,10 +325,10 @@
     
-    <style name="CustomButtonStyleOutline" parent="Widget.MaterialComponents.Button.OutlinedButton">
-        <item name="shapeAppearanceOverlay">@style/RoundedButtonStyle</item>
+    <style name="NegativeButtonStyle" parent="Widget.MaterialComponents.Button.OutlinedButton">
```
@shubham1g5 I think this is more appropriate

avazirna commented:
@damagatchi retest this please

avazirna commented:
@coderabbitai full review

avazirna commented on app/res/drawable/button_orange_background.xml:
```diff
@@ -1,8 +1,24 @@
+        <solid android:color="@color/orange_500" />
+        <corners android:radius="22dp" />
+    </shape>
+</item>
```
@shubham1g5 I have reverted this as it didn't work as expected, some posts suggest that [direct reference to a theme attribute from a state selector doesn't work](https://issuetracker.google.com/issues/36941443#comment25), maybe in older versions of Android. Also the _selectors_ where defined for `AppCompatButton` not `MaterialButton`, I have adjusted the existing styles [9e87ddb](https://github.com/dimagi/commcare-android/pull/2897/commits/9e87ddbd52b78c700f890b38d3715f2e45637cf7)

avazirna commented:
@shubham1g5 @pm-dimagi I believe we can merge this. I did a pass around the aspects that `coderabbit` flagged and I didn't see anything major, can you both do the same?

avazirna commented:
@damagatchi retest this please
